untyped-lambda-calculus/article.md

```diff
@@ -0,0 +1,508 @@
+>
+>  -- Felice Cardone, J. Roger Hindley, History of Lambda-calculus and Combinatory Logic
+
+After this, admittedly, a very brief look at the untyped lambda calculus, you hopefully have a bit of a feel for how it works.
+
+To summarize:
+
+- Lambda calculus is a theoretical framework on which functional programming languages are built.
+- Lambda calculus has two kinds: untyped and typed.
+- The pure untyped lambda calculus only consists of variables, abstractions (i.e. univariate functions), and function applications.
+- We can represent any computable function in pure untyped lambda calculus.
+- Various encodings can be used to represent booleans, naturals, tuples, etc.
+
+The last three points are somewhat profound if you think about it: we only defined a single operation (two if you count the definition of lambda abstraction), and that got us to anything theoretically computable.
+There are other constructions of the universal computer: Turing and Post machines, the Markov algorithm, and general recursive functions.
+But, arguably, nothing beats the simplicity of lambda calculus.
+One has to appreciate the elegance at least.
+
+The primary point of interest is, of course, not so much the calculus itself but its applications, of which there are many.
+The untyped kind, however, doesn't find much use beyond the computability theory.
+Still, it's necessary to understand the untyped lambda calculus to meaningfully discuss the typed variety (or, more accurately, varieties).
+
+As for more practical applications, here's a few languages based on lambda calculus:
+
+- Haskell;
+- most languages in the ML family, like OCaml and F#;
+- most dependently typed languages, like Agda, Coq, F*, Idris;
+- Cardano blockchain's Plutus Core.
+
+## Exercises
+
+1. We introduced the implementation of the boolean conjunction for Church-encoded boolean values (the `and` operation).
+    Implement disjunction (`or`), inversion (`not`), and exclusive disjunction (`xor`).
+
+2. We did introduce one particular implementation for the successor function for Church-encoded naturals, `succ`.
+    It is not the only possible implementation, however.
+    Suggest another possible implementation, which is not alpha-equivalent to the one we introduced.
+
+3. Define an exponentiation function for Church-encoded naturals.
+
+4. Define subtraction of Church-encoded naturals using the predecessor function `pred`.
+    For the case when the difference is undefined in naturals, i.e. when the subtrahend is bigger than minuend, feel free to return $c_0$ or $\Omega\;\Omega$.
+
+5. Define a function that takes two Church-encoded naturals and returns `tru` if those are equal and `fls` otherwise.
+
+6. Implement a pure untyped lambda calculus interpreter in your favorite programming language.
+
+<p>
+<details>
+<summary>Tips on handling name collisions</summary>
```

Comment:
Unfortunately, this doesn't go with the list, I assume there needs to be some indentation. 🤔

<img width="526" alt="image" src="https://user-images.githubusercontent.com/35335196/165775232-6862605b-8440-491e-9552-8f9b4eb4c07f.png">

Reply:
```
4. As you might remember, a semigroup is a monoid without the identity element. Can you come up with an example of a structure that is a semigroup but is not a monoid? 

    <details>
      <summary>Solution</summary>
    
	One example is [non-empty lists](https://hackage.haskell.org/package/base-4.16.0.0/docs/Data-List-NonEmpty.html) with the `++` operator. While a regular list can be empty, and the empty list thus satisfies the requirement of an identity element, non-empty lists have no such element.
	
    <hr>
    </details>
```

Here's what worked for my monoid article, if that helps.